analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion.cs

```diff
@@ -18,32 +18,43 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-using System;
-using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Linq;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Diagnostics;
 using SonarAnalyzer.Common;
-using SonarAnalyzer.CFG.Sonar;
 using SonarAnalyzer.Helpers;
 using StyleCop.Analyzers.Lightup;
 
 namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     [Rule(DiagnosticId)]
-    public sealed class InfiniteRecursion : SonarDiagnosticAnalyzer
+    public class InfiniteRecursion : SonarDiagnosticAnalyzer
     {
-        internal const string DiagnosticId = "S2190";
+        private const string DiagnosticId = "S2190";
         private const string MessageFormat = "Add a way to break out of this {0}.";
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private readonly IInfiniteRecursion analyzer;
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        public static DiagnosticDescriptor Rule => DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public InfiniteRecursion() : this(false) { }
+
+        public InfiniteRecursion(bool runSonarAnalyzer)
+        {
+            if (!CFG.Roslyn.ControlFlowGraph.IsAvailable || runSonarAnalyzer)
+            {
+                analyzer = new InfiniteRecursionSonarCfg();
+            }
+            else
+            {
+                analyzer = new InfiniteRecursionRoslynCfg();
+            }
```

Comment:
Why not shorten this down to:

```suggestion
            analyzer = !CFG.Roslyn.ControlFlowGraph.IsAvailable || runSonarAnalyzer ? new InfiniteRecursionSonarCfg() : new InfiniteRecursionRoslynCfg();
```
?

Reply:
Good point. Please note that this is a draft, that means that I am just making it work and do no not clean the code yet, so code contains some duplicated/uncleaned code, since we want to verify what works and what does not.